common/models/templates/src/tokenization.ts

```diff
@@ -22,21 +28,66 @@ export interface Tokenization {
   caretSplitsToken: boolean
 }
 
-export function tokenize(wordBreaker: WordBreakingFunction, context?: Partial<Context>): Tokenization {
+export function tokenize(
+  wordBreaker: WordBreakingFunction,
+  context: Partial<Context>,
+  options?: {
+    /** Characters to rejoin to preceding tokens if found immediately pre-caret. */
+    rejoins?: string[]
+  }
+): Tokenization {
+  const rejoins = options?.rejoins || ["'"];
   context = context || {
     left: undefined,
     startOfBuffer: undefined,
     endOfBuffer: undefined
   };
 
-  let leftSpans  = wordBreaker(context.left || '') || [];
-  let rightSpans = wordBreaker(context.right || '') || [];
+  const leftSpans  = wordBreaker(context.left  || '') || [];
+  const rightSpans = wordBreaker(context.right || '') || [];
+
+  const tokenization: Tokenization = {
+    left: [],
+    right: [],
+    caretSplitsToken: false
+  }
 
-  let leftTail: Span;
-  if(leftSpans.length > 0) {
-    leftTail = leftSpans[leftSpans.length - 1];
+  // New step 1:  process left-context.
+  let currentIndex = 0;
+  while(leftSpans.length > 0) {
+    const nextSpan = leftSpans[0];
+    if(nextSpan.start != currentIndex) {
+      // Implicit whitespace span!
+      tokenization.left.push({
+        text: context.left!.substring(currentIndex, nextSpan.start),
+        isWhitespace: true
+      });
+      currentIndex = nextSpan.start;
+    } else {
+      leftSpans.shift();
+      // Explicit non-whitespace span.
+      tokenization.left.push({
+        text: nextSpan.text
+      });
+      currentIndex = nextSpan.end;
+    }
   }
 
+  // Detect any pre-caret whitespace after the final pre-caret non-whitespace
+  // token
+  //
+  // Note:  the default wordbreaker won't need this code, as it emits a `''`
+  // after final whitespace.
+  if(currentIndex != (context.left?.length ?? 0)) {
+    tokenization.left.push({
+      text: context.left!.substring(currentIndex, context.left!.length),
+      isWhitespace: true
+    });
+    currentIndex = context.left!.length;
+  }
```

Comment:
This seems to be an incorrect use of the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) (`!`) on line 83 -- that operator is asserting that you know `context.left` cannot be nullish -- but the test on line 81 allows a nullish `context.left`.

I am not entirely sure what you are intending; is this right?

```suggestion
  if(context.left != null && currentIndex != context.left.length) {
    tokenization.left.push({
      text: context.left.substring(currentIndex, context.left.length),
      isWhitespace: true
    });
    currentIndex = context.left.length;
  }
```

Yes, I can see how it could work out the same, if `currentIndex == 0`, but I am not confident from reading the code that that would always be the case. If in doubt, make the code simpler!